### ISSUE TITLE
Something wrong with F.cross_entropy, it is different from average cross entropy of all categories

### ISSUE THREAD

**Issue by xinliweiyuan** [pytorch/pytorch#22957]
## To Reproduce

```
targets = torch.LongTensor([3, 0, 2, 0, 1, 0, 2, 3, 0, 2, 6, 1, 1, 4, 1, 3, 3, 0, 1, 5, 1, 2, 0, 1])
inputs = torch.FloatTensor([[ -4.3700,  -3.9679,  -3.0123,  -1.7961,  -0.6880,  -1.7625,  -2.9644,
          -4.0502,  -4.5962],
        [ -0.8429,  -0.9089,  -2.1807,  -3.1729,  -4.6232,  -6.9062,  -7.4340,
          -8.5912,  -8.8176],
        [ -5.6330,  -3.4058,  -0.1488,  -2.4443,  -4.3518,  -6.7693,  -7.9675,
          -8.6363,  -8.7170],
        [ -0.7802,  -0.9133,  -2.4122,  -3.3046,  -4.4443,  -6.5722,  -7.2875,
          -8.5103,  -8.8142],
        [ -0.5500,  -1.0661,  -2.8024,  -4.3089,  -5.7051,  -7.4434,  -7.8173,
          -8.7027,  -8.9205],
        [ -0.1407,  -2.1787,  -4.1899,  -6.1222,  -7.6477,  -9.4781,  -9.4218,
         -10.2963, -10.5539],
        [ -3.4817,  -1.8641,  -1.0104,  -1.1131,  -2.2736,  -4.2750,  -5.8090,
          -6.7200,  -7.4932],
        [ -1.4598,  -0.6211,  -1.9367,  -2.6544,  -4.3054,  -6.5809,  -7.4238,
          -8.4327,  -8.6501],
        [ -1.3426,  -1.0558,  -1.4547,  -2.1734,  -3.3196,  -5.3505,  -6.3363,
          -7.4134,  -7.8894],
        [ -2.6378,  -2.2429,  -1.5336,  -1.2713,  -1.3751,  -2.9832,  -4.1769,
          -5.2808,  -6.1078],
        [ -2.8716,  -4.4096,  -4.2283,  -4.1608,  -3.3809,  -2.8343,  -2.3552,
          -1.7596,  -0.6138],
        [ -2.3647,  -0.6851,  -1.1531,  -2.6243,  -4.4309,  -6.7563,  -7.7475,
          -8.5107,  -8.7392],
        [ -2.4920,  -1.3769,  -0.9216,  -1.6013,  -2.9389,  -4.7561,  -6.0067,
          -7.0142,  -7.6498],
        [ -1.5442,  -3.0835,  -2.9438,  -2.9264,  -1.7392,  -1.9187,  -2.1094,
          -2.5482,  -2.1849],
        [ -0.7196,  -0.8927,  -2.5220,  -3.9956,  -5.5137,  -7.7908,  -8.4999,
          -9.4648,  -9.4958],
        [ -4.0392,  -1.4743,  -0.7293,  -1.4813,  -3.2523,  -5.5743,  -7.1149,
          -7.8801,  -8.6374],
        [ -1.7170,  -0.8503,  -1.2964,  -2.3758,  -3.8023,  -5.8127,  -7.1644,
          -8.0118,  -8.3362],
        [ -0.4389,  -1.4297,  -2.7526,  -3.3089,  -4.3297,  -6.5700,  -7.3215,
          -8.4688,  -8.9631],
        [ -1.3658,  -0.6943,  -1.6698,  -3.0700,  -4.7176,  -6.9264,  -7.7186,
          -8.5074,  -8.8209],
        [ -4.6864,  -3.0129,  -1.6946,  -0.6187,  -1.6444,  -3.9086,  -5.4969,
          -6.6174,  -7.3285],
        [ -3.4751,  -0.2108,  -1.9760,  -4.0342,  -6.0920,  -8.1109,  -8.9523,
          -9.5838,  -9.8464],
        [ -2.0389,  -0.8592,  -1.0708,  -2.4411,  -4.2490,  -6.5746,  -7.6327,
          -8.4447,  -8.6728],
        [ -0.6506,  -0.9062,  -2.8058,  -4.5070,  -6.2096,  -8.1341,  -8.3638,
          -9.4096,  -9.6140],
        [ -1.4527,  -0.4381,  -2.2409,  -4.4155,  -6.3280,  -8.2989,  -8.6995,
          -9.4820,  -9.5856]])


loss = 0
N = torch.max(targets)
for i in range(N):
    logits_mask = inputs[targets==i]
    targets_mask = targets[targets==i]
    loss_i = F.cross_entropy(logits_mask, targets_mask)
    loss += loss_i
    
print(loss/N)

print(F.cross_entropy(inputs, targets))
```

results:
tensor(1.6885)
tensor(1.2348)

Why they are different? Can someone explain this, thanks! Moreover, if I want to calculate the loss of different categories, is there a better solution rather than iterations.

**Comment by xinliweiyuan** [comment]
also when we give weights (all less than 1.0) to each class, the loss should be less than total cross-entropy, while it is much larger than the total entropy.
print(F.cross_entropy(inputs, targets, weight=weights))
tensor(1.5703)

**Comment by Jonas1312** [comment]
```
loss = 0
N = targets.size(0)
for i in range(torch.max(targets)+1):
    logits_mask = inputs[targets==i]
    targets_mask = targets[targets==i]
    loss_i = F.cross_entropy(logits_mask, targets_mask, reduction="sum")
    loss += loss_i
    
print(loss/N)

print(F.cross_entropy(inputs, targets))
```

**Comment by Jonas1312** [comment]
I highly recommend to use weights that sum to 1.0 in order to get the same loss amplitude:

```
    weights = torch.FloatTensor([1.0 / x for x in weights]).to(device)
    weights = weights / torch.sum(weights)
    print("Class weights: ", weights)  # should sum to 1
```

**Comment by xinliweiyuan** [comment]
Thanks for correcting. But the loss still different from the sum of all the categories.

**Comment by Jonas1312** [comment]
```
import torch
from torch.nn import functional as F

targets = torch.LongTensor([3, 0, 2, 0, 1, 0, 2, 3, 0, 2, 6, 1, 1, 4, 1, 3, 3, 0, 1, 5, 1, 2, 0, 1])
inputs = torch.FloatTensor([[ -4.3700,  -3.9679,  -3.0123,  -1.7961,  -0.6880,  -1.7625,  -2.9644,
          -4.0502,  -4.5962],
        [ -0.8429,  -0.9089,  -2.1807,  -3.1729,  -4.6232,  -6.9062,  -7.4340,
          -8.5912,  -8.8176],
        [ -5.6330,  -3.4058,  -0.1488,  -2.4443,  -4.3518,  -6.7693,  -7.9675,
          -8.6363,  -8.7170],
        [ -0.7802,  -0.9133,  -2.4122,  -3.3046,  -4.4443,  -6.5722,  -7.2875,
          -8.5103,  -8.8142],
        [ -0.5500,  -1.0661,  -2.8024,  -4.3089,  -5.7051,  -7.4434,  -7.8173,
          -8.7027,  -8.9205],
        [ -0.1407,  -2.1787,  -4.1899,  -6.1222,  -7.6477,  -9.4781,  -9.4218,
         -10.2963, -10.5539],
        [ -3.4817,  -1.8641,  -1.0104,  -1.1131,  -2.2736,  -4.2750,  -5.8090,
          -6.7200,  -7.4932],
        [ -1.4598,  -0.6211,  -1.9367,  -2.6544,  -4.3054,  -6.5809,  -7.4238,
          -8.4327,  -8.6501],
        [ -1.3426,  -1.0558,  -1.4547,  -2.1734,  -3.3196,  -5.3505,  -6.3363,
          -7.4134,  -7.8894],
        [ -2.6378,  -2.2429,  -1.5336,  -1.2713,  -1.3751,  -2.9832,  -4.1769,
          -5.2808,  -6.1078],
        [ -2.8716,  -4.4096,  -4.2283,  -4.1608,  -3.3809,  -2.8343,  -2.3552,
          -1.7596,  -0.6138],
        [ -2.3647,  -0.6851,  -1.1531,  -2.6243,  -4.4309,  -6.7563,  -7.7475,
          -8.5107,  -8.7392],
        [ -2.4920,  -1.3769,  -0.9216,  -1.6013,  -2.9389,  -4.7561,  -6.0067,
          -7.0142,  -7.6498],
        [ -1.5442,  -3.0835,  -2.9438,  -2.9264,  -1.7392,  -1.9187,  -2.1094,
          -2.5482,  -2.1849],
        [ -0.7196,  -0.8927,  -2.5220,  -3.9956,  -5.5137,  -7.7908,  -8.4999,
          -9.4648,  -9.4958],
        [ -4.0392,  -1.4743,  -0.7293,  -1.4813,  -3.2523,  -5.5743,  -7.1149,
          -7.8801,  -8.6374],
        [ -1.7170,  -0.8503,  -1.2964,  -2.3758,  -3.8023,  -5.8127,  -7.1644,
          -8.0118,  -8.3362],
        [ -0.4389,  -1.4297,  -2.7526,  -3.3089,  -4.3297,  -6.5700,  -7.3215,
          -8.4688,  -8.9631],
        [ -1.3658,  -0.6943,  -1.6698,  -3.0700,  -4.7176,  -6.9264,  -7.7186,
          -8.5074,  -8.8209],
        [ -4.6864,  -3.0129,  -1.6946,  -0.6187,  -1.6444,  -3.9086,  -5.4969,
          -6.6174,  -7.3285],
        [ -3.4751,  -0.2108,  -1.9760,  -4.0342,  -6.0920,  -8.1109,  -8.9523,
          -9.5838,  -9.8464],
        [ -2.0389,  -0.8592,  -1.0708,  -2.4411,  -4.2490,  -6.5746,  -7.6327,
          -8.4447,  -8.6728],
        [ -0.6506,  -0.9062,  -2.8058,  -4.5070,  -6.2096,  -8.1341,  -8.3638,
          -9.4096,  -9.6140],
        [ -1.4527,  -0.4381,  -2.2409,  -4.4155,  -6.3280,  -8.2989,  -8.6995,
          -9.4820,  -9.5856]])


loss = 0
N = targets.size(0)
for i in range(torch.max(targets)+1):
    logits_mask = inputs[targets==i]
    targets_mask = targets[targets==i]
    loss_i = F.cross_entropy(logits_mask, targets_mask, reduction="sum")
    loss += loss_i
    
print(loss/N)

print(F.cross_entropy(inputs, targets))
```

gives:

```
tensor(1.2348)
tensor(1.2348)
```

**Comment by xinliweiyuan** [comment]
But the document says, \text{loss}(x, class) = weight[class] \left(-x[class] + \log\left(\sum_j \exp(x[j])\right)\right). My weights is = torch.FloatTensor([0.5, 0.5, 0.7, 0.8, 0.9, 1.0, 1.5, 2.0, 2.0]). It is true, we should normalize the weigths. Actually I want to address the unbalance classification problem by giving different weights. Do you have any other suggestions? Thanks. I got to think about it.

**Comment by Jonas1312** [comment]
Sorry I don't get it

If you're struggling with class imbalance you should have a look at focal loss maybe

**Comment by xinliweiyuan** [comment]
Yep, I tried it, but not worked well. Thanks for pointing it out.

**Comment by fmassa** [comment]
The reason for the difference has been posted by @Jonas1312 in https://github.com/pytorch/pytorch/issues/22957#issuecomment-512224779

If you have modeling issues, please use the forums for questions https://discuss.pytorch.org
